commands.js

```diff
@@ -365,138 +360,7 @@ class Commands {
-            badges += `:hammer: I like trying to break Pinano Bot\n`
-            break
-        }
-      }
-
-      if (user.quiz_score >= 10) {
-        badges += `:question: I've correctly answered ${user.quiz_score} riddles on [#🎶literature-quiz]` +
-          '(https://discordapp.com/channels/188345759408717825/505872476903964674)\n'
-      }
-
-      if (user.riddles_solved >= 10) {
-        badges += `:question: My riddles have been solved ${user.riddles_solved} times on [#🎶literature-quiz]` +
-          '(https://discordapp.com/channels/188345759408717825/505872476903964674)\n'
-      }
-
-      if (user.recitals != null && user.recitals.length >= 3) {
-        // one for each note in the emoji
-        badges += `:notes: I've played in ${user.recitals.length} recitals\n`
-      }
-
-      if (hasLongSession || now - user.last_practiced_time < 7 * 86400) {
-        badges += ':calendar: I\'ve practiced within the last week\n'
-      } else if (now - user.last_practiced_time < 30 * 86400) {
-        badges += ':calendar: I\'ve practiced within the last thirty days\n'
-      }
-    }
-
-    if (now * 1000 - mem.joinedTimestamp >= 88 * 86400 * 1000) {
-      // join date is more than 88 days ago
-      badges += ':musical_keyboard: I joined Pinano at least 88 days ago\n'
-    }
-
-    if (mem.roles.some(r => r.name === 'Hand Revealed')) {
-      badges += ':hand_splayed: I\'ve revealed my hand on [#hand-reveals](https://discordapp.com/channels/188345759408717825/440705391454584834)\n'
-    }
-
-    if (settings.contributors.includes(mem.id)) {
-      badges += ':robot: I\'ve contributed code to Pinano Bot on [GitHub](https://github.com/pinano-discord/Pinano-Discord-Bot)\n'
-    }
-
-    if (userInfo.overallSession >= 500 * 60 * 60) {
-      badges += '<:FiveHundredHours:627099475701268480> I\'ve practiced for at least 500 hours\n'
-    } else if (userInfo.overallSession >= 250 * 60 * 60) {
-      badges += '<:TwoFiftyHours:627099476120829982> I\'ve practiced for at least 250 hours\n'
-    } else if (userInfo.overallSession >= 100 * 60 * 60) {
-      badges += '<:HundredHours:627099476078755850> I\'ve practiced for at least 100 hours\n'
-    } else if (userInfo.overallSession >= 40 * 60 * 60) {
-      badges += '<:FortyHours:627099475869171712> I\'ve practiced for at least 40 hours\n'
-    }
-
-    let effectiveName = ((mem.nickname == null) ? mem.user.username : mem.nickname).toLowerCase()
-    if (effectiveName.endsWith('juice') || effectiveName.endsWith('juwuice')) {
-      badges += ':tropical_drink: I miss minijuice\n'
-    }
-
-    // thanks for being an awesome tester, Sayee
-    if (userInfo.mem.id === '151301657915817984') {
-      badges += `:hammer: I like trying to break Pinano Bot\n`
-    }
-
-    if (badges === '') {
-      badges = '<:wtf:593197993264414751> no badges yet!'
-    }
-
-    embed.addField('Badges', badges)
+      .addField('Badges', badgesForUser(userInfo, user, hasLongSession))
```

Comment:
Passing all this stuff in here feels a little gross to me, but it is expedient.

I think longer term, I might imagine putting everything into `userInfo`, making that more of a well-defined struct as opposed to an object with a bunch of properties.

Reply:
I think there are a couple of ways we could do badges.

One is for a badge to be a tuple of a `(lambda, displayText, supercedes)`. Call each lambda on a `user` or `userInfo` (or both), wipe out everything that's superceded by another badge that also tests positive, return all the `displayText`s. Existing updates to stats checked by badges still happen in the same place.

Another way is for badges to be actually part of the user record. When the relevant stat updates, award the badge to the user if it's not already awarded.

Reply:
If I were using SQL (or hbase or some such), I'd probably lean towards having a badge table that is 1 user : many badges. Perhaps each row being `(userId, timestamp, badge_type, display_text)` (and guild??).

Then for supercedes, perhaps each badge has an "type" and somehow approximate `group by badge_type` and select the latest badge of each type.

Could have a job that populates that table from state in the user record.